### PR TITLE
Hazards: Turret, Explosive Barrel, Floor Trap. Also, Explosion.

### PR DESCRIPTION
**Pull request by edjohn**
Turret: Periodically shoots energy projectiles at a player within its radius. Explodes on death.
Explosive Barrel: Explodes on death.
Floor Trap: Consistently damages the player while they're standing on it. Doesn't die.
Explosion: Deals damage to all enemies, players, and destructibles in its explosion radius.

**Important**: HealthSystem had to be revised to have a nullcheck for the takeDamage function. It will break all scripts that utilize its damage functionality if it doesn't.

### CODE REVIEW

File: Assets/Scripts/Hazards/EnergyProjectile.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            Explode();
+        }
+        // Move towards the target by a step value every frame.
+        float step = _speed * Time.deltaTime;
+        transform.position = Vector3.MoveTowards(transform.position, _targetPosition, step);
+    }
+
+    // Required initialization for projectile to have a target to move towards.
+    public void InitializeWithTargetPos(Vector3 pos)
```
Comment (dominicaq) [inline Assets/Scripts/Hazards/EnergyProjectile.cs]: Can you rename this to `SetTargetPosition(Vector3 position)`

File: Assets/Scripts/Hazards/ExplosiveBarrel.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using Unity.Netcode;
+using Shared;
+
+public class ExplosiveBarrel : NetworkBehaviour
+{
+    private HealthSystem _health;
+    // Prefab for explosion that will be spawned on turret death.
+    [SerializeField] private GameObject _explosionPrefab;
+
+    void Awake()
```
Comment (dominicaq) [inline Assets/Scripts/Hazards/ExplosiveBarrel.cs]: For consistency can you make this `private void Start()`

File: Assets/Scripts/Hazards/Explosion.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using Shared;
+
+public class Explosion : NetworkBehaviour
+{
+    [SerializeField] private float _explosionDamage;
+    [SerializeField] private float _explosionRadius = 2f;
+    private const float _explosionDuration = 2f;
+    private AudioSource _explosionAudio;
+
+    void Awake()
```
Comment (dominicaq) [inline Assets/Scripts/Hazards/Explosion.cs]: Same here, can you make this `private void Start()`?

File: Assets/Scripts/Environment/Turret.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    [SerializeField] private float _damage;
+    [SerializeField] private float _detectionRadius;
+    [SerializeField] private float _attackCooldown;
+    private float _cooldownProgress;
+
+    // Throttles polling for targets in detection radius. No reason for turret to reorient itself on every single frame.
+    private float _timeToNextTargetCheck;
+    private float _targetCheckTime = 0.05f;
+
+    // Prefab for explosion that will be spawned on turret death.
```
Comment (dominicaq) [inline Assets/Scripts/Environment/Turret.cs:19]: Remove redundant comments, we can read types and variable names lol

File: Assets/Scripts/Environment/Turret.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                if (!_audio.isPlaying)
+                {
+                    _audio.Play();
+                }
+                transform.LookAt(collider.transform);
+                return collider.gameObject;
+            }
+        }
+        _audio.Stop();
```
Comment (dominicaq) [inline Assets/Scripts/Environment/Turret.cs]: Won`t this stop the audio from playing once you find the target? Seems you're trying to play audio when you find a target which is cool.

**Review by dominicaq — COMMENTED** [review]
Looks good, most of my comments are just renaming some stuff. For code consistency make all of your `Awake()` functions `private void Start()` unless you need awake specifically. Can you remove the `Dungeon Floor Traps` folder and move its contents into our `Meshes` `Texture` and `Material` folders? Just keeping the project as clean as possible (IMO it's already a little messy). Once you do the small clean up you should be all good.

**Comment by dominicaq** [comment]
image
When the turret dies make sure to disable the base of the turret's collider so you don't block the player or allow them to jump on top of the invisible box.

File: Assets/Scripts/Environment/Turret.cs (hunk 5 above)
Comment (edjohn) [reply Assets/Scripts/Environment/Turret.cs]: The turret should play a "detection" audio as long as a player is in its range. By stopping the audio source, the turret doesn't keep playing the audio even after the player is out of range.

**Review by dominicaq — APPROVED** [review]
Looks good

**Pull request merged**